File: go/internal/platforms/zai_tts/tts.go
// Package zai_tts implements the Z.ai TTS (Text-to-Speech) API client.
// Endpoint: POST https://audio.z.ai/api/v1/z-audio/tts/create
// Auth: Bearer JWT token.
// Response: SSE stream with {"audio":"<base64 WAV>"} chunks, ending with [DONE].
package zai_tts

import (
	"bufio"
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

const (
	DefaultEndpoint     = "https://audio.z.ai/api/v1/z-audio/tts/create"
	DefaultAuthEndpoint = "https://audio.z.ai/api/v1/z-audio/auth/"
	DefaultTimeout      = 120 * time.Second
)

// TTSClient calls the Z.ai TTS API.
type TTSClient struct {
	Endpoint     string
	AuthEndpoint string
	Token        string
	UserID       string
	Client       *http.Client
}

// NewTTSClient creates a client with the given Bearer token and user ID.
func NewTTSClient(token, userID string) *TTSClient {
	return &TTSClient{
		Endpoint:     DefaultEndpoint,
		AuthEndpoint: DefaultAuthEndpoint,
		Token:        token,
		UserID:       userID,
		Client:       &http.Client{Timeout: DefaultTimeout},
	}
}

// Authenticate exchanges an OAuth code for a token. Auto-sets Token and UserID.
func (c *TTSClient) Authenticate(code string) (*AuthResponse, error) {
	payload, _ := json.Marshal(AuthRequest{Code: code})
	req, err := http.NewRequest(http.MethodPost, c.AuthEndpoint, bytes.NewReader(payload))
	if err != nil {
		return nil, fmt.Errorf("create auth request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "*/*")
	req.Header.Set("Origin", "https://audio.z.ai")
	req.Header.Set("Referer", "https://audio.z.ai/")

	resp, err := c.Client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("auth request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read auth response: %w", err)
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("auth HTTP %d: %s", resp.StatusCode, string(body))
	}

	var result AuthResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("parse auth response: %w", err)
	}
	if result.Data.AuthToken != "" {
		c.Token = result.Data.AuthToken
	}
	if result.Data.UserID != "" {
		c.UserID = result.Data.UserID
	}
	return &result, nil
}

// Synthesize converts text to speech and returns the complete WAV audio bytes.
// It reads the SSE stream, collects all base64 audio chunks, and concatenates them.
func (c *TTSClient) Synthesize(req TTSRequest) ([]byte, error) {
	if req.UserID == "" {
		req.UserID = c.UserID
	}
	if req.Speed == 0 {
		req.Speed = 1
	}
	if req.Volume == 0 {
		req.Volume = 1
	}

	payload, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("marshal request: %w", err)
	}

	httpReq, err := http.NewRequest(http.MethodPost, c.Endpoint, bytes.NewReader(payload))
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("Accept", "text/event-stream")
	httpReq.Header.Set("Authorization", "Bearer "+c.Token)
	httpReq.Header.Set("Origin", "https://audio.z.ai")
	httpReq.Header.Set("Referer", "https://audio.z.ai/")

	resp, err := c.Client.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("http request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body))
	}

	return readSSEAudio(resp.Body)
}

// SynthesizeSimple is a convenience method with default voice settings.
func (c *TTSClient) SynthesizeSimple(text string) ([]byte, error) {
	return c.Synthesize(TTSRequest{
		VoiceName: "通用男声",
		VoiceID:   "system_003",
		InputText: text,
		Speed:     1,
		Volume:    1,
	})
}

// readSSEAudio reads SSE events and collects base64 audio chunks into raw bytes.
func readSSEAudio(r io.Reader) ([]byte, error) {
	var audioData []byte
	scanner := bufio.NewScanner(r)
	scanner.Buffer(make([]byte, 0, 4*1024*1024), 16*1024*1024) // up to 16MB per line

	for scanner.Scan() {
		line := scanner.Text()
		if !strings.HasPrefix(line, "data: ") {
			continue
		}
		data := strings.TrimPrefix(line, "data: ")
		if data == "[DONE]" {
			break
		}

		var chunk struct {
			Audio string `json:"audio"`
		}
		if err := json.Unmarshal([]byte(data), &chunk); err != nil {
			continue
		}
		if chunk.Audio == "" {
			continue
		}

		decoded, err := base64.StdEncoding.DecodeString(chunk.Audio)
		if err != nil {
			continue
		}
		audioData = append(audioData, decoded...)
	}

	if err := scanner.Err(); err != nil {
		return audioData, fmt.Errorf("read SSE stream: %w", err)
	}
	return audioData, nil
}

File: python/providers/zai_tts/tts.py
"""Z.ai TTS API client — converts text to speech via SSE audio stream."""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass, field
from typing import Any
from urllib.request import Request, urlopen


DEFAULT_ENDPOINT = "https://audio.z.ai/api/v1/z-audio/tts/create"
DEFAULT_AUTH_ENDPOINT = "https://audio.z.ai/api/v1/z-audio/auth/"
DEFAULT_TIMEOUT = 120


@dataclass
class AuthResponse:
    """Parsed auth API response."""
    code: int = 0
    message: str = ""
    user_id: str = ""
    auth_token: str = ""
    name: str = ""
    profile_image_url: str = ""
    timestamp: int = 0
    raw: dict[str, Any] = field(default_factory=dict)


class TTSClient:
    """Client for the Z.ai TTS API."""

    def __init__(self, token: str = "", user_id: str = "",
                 endpoint: str = DEFAULT_ENDPOINT,
                 auth_endpoint: str = DEFAULT_AUTH_ENDPOINT,
                 timeout: int = DEFAULT_TIMEOUT):
        self.token = token
        self.user_id = user_id
        self.endpoint = endpoint
        self.auth_endpoint = auth_endpoint
        self.timeout = timeout

    def authenticate(self, code: str) -> AuthResponse:
        """Exchange an OAuth code for a token. Auto-sets token and user_id."""
        payload = json.dumps({"code": code}).encode("utf-8")
        req = Request(self.auth_endpoint, data=payload, method="POST")
        req.add_header("Content-Type", "application/json")
        req.add_header("Accept", "*/*")
        req.add_header("Origin", "https://audio.z.ai")
        req.add_header("Referer", "https://audio.z.ai/")

        with urlopen(req, timeout=self.timeout) as resp:
            raw = json.loads(resp.read().decode("utf-8"))

        d = raw.get("data", {}) if isinstance(raw.get("data"), dict) else {}
        result = AuthResponse(
            code=raw.get("code", 0), message=raw.get("message", ""),
            user_id=d.get("user_id", ""), auth_token=d.get("auth_token", ""),
            name=d.get("name", ""), profile_image_url=d.get("profile_image_url", ""),
            timestamp=raw.get("timestamp", 0), raw=raw,
        )
        if result.auth_token:
            self.token = result.auth_token
        if result.user_id:
            self.user_id = result.user_id
        return result

    @classmethod
    def from_code(cls, code: str, **kwargs: Any) -> tuple[TTSClient, AuthResponse]:
        """Create a client by authenticating with an OAuth code."""
        client = cls(**kwargs)
        auth = client.authenticate(code)
        return client, auth

    def synthesize(self, text: str, voice_id: str = "system_003",
                   voice_name: str = "通用男声", speed: float = 1,
                   volume: float = 1) -> bytes:
        """Convert text to speech. Returns WAV audio bytes."""
        payload = json.dumps({
            "voice_name": voice_name, "voice_id": voice_id,
            "user_id": self.user_id, "input_text": text,
            "speed": speed, "volume": volume,
        }).encode("utf-8")

        req = Request(self.endpoint, data=payload, method="POST")
        req.add_header("Content-Type", "application/json")
        req.add_header("Accept", "text/event-stream")
        req.add_header("Authorization", f"Bearer {self.token}")
        req.add_header("Origin", "https://audio.z.ai")
        req.add_header("Referer", "https://audio.z.ai/")

        with urlopen(req, timeout=self.timeout) as resp:
            return self._read_sse_audio(resp)

    @staticmethod
    def _read_sse_audio(resp: Any) -> bytes:
        """Read SSE events and collect base64 audio chunks into raw bytes."""
        audio_data = bytearray()
        for raw_line in resp:
            line = raw_line.decode("utf-8", errors="replace").strip()
            if not line.startswith("data: "):
                continue
            data = line[6:]
            if data == "[DONE]":
                break
            try:
                chunk = json.loads(data)
            except json.JSONDecodeError:
                continue
            audio_b64 = chunk.get("audio", "")
            if audio_b64:
                audio_data.extend(base64.b64decode(audio_b64))
        return bytes(audio_data)
